cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeEngineCredentials.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.runtime.spi.provisioner.dataproc;
+
+import com.google.api.client.util.ExponentialBackOff;
+import com.google.api.client.util.GenericData;
+import com.google.auth.oauth2.AccessToken;
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.common.io.CharStreams;
+import com.google.gson.Gson;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Date;
+
+/**
+ * Provides ComputeEngineCredentials either locally if no endpoint is defined, or remotely if endpoint is defined.
+ */
+public final class ComputeEngineCredentials extends GoogleCredentials {
+  private static final Logger LOG = LoggerFactory.getLogger(ComputeEngineCredentials.class);
+  private static final Gson GSON = new Gson();
+  /**
+   * Time (in millisecond) to refresh the credentials before it expires.
+   */
+  private static final int NUMBER_OF_RETRIES = 10;
+  private static final int MIN_WAIT_TIME = 5000;
```

Comment:
min_wait_time of 5 seconds is too high? maybe 100ms

Reply:
IMO, 100ms is too short since it takes roughly 30sec for worker pod to become available. Will change it to 500 ms.